lib/IDE/Refactoring.cpp

```diff
@@ -4545,32 +4545,135 @@ struct CallbackCondition {
   }
 };
 
+/// A list of nodes to print, along with a list of locations that may have
+/// preceding comments attached, which also need printing. For example:
+///
+/// \code
+/// if .random() {
+///   // a
+///   print("hello")
+///   // b
+/// }
+/// \endcode
+///
+/// To print out the contents of the if statement body, we'll include the AST
+/// node for the \c print call. This will also include the preceding comment
+/// \c a, but won't include the comment \c b. To ensure the comment \c b gets
+/// printed, the SourceLoc for the closing brace \c } is added as a possible
+/// comment loc.
+class NodesToPrint {
+  SmallVector<ASTNode, 0> Nodes;
+  SmallVector<SourceLoc, 2> PossibleCommentLocs;
+
+public:
+  NodesToPrint() {}
+  NodesToPrint(ArrayRef<ASTNode> Nodes, ArrayRef<SourceLoc> PossibleCommentLocs)
+      : Nodes(Nodes.begin(), Nodes.end()),
+        PossibleCommentLocs(PossibleCommentLocs.begin(),
+                            PossibleCommentLocs.end()) {}
+
+  ArrayRef<ASTNode> getNodes() const { return Nodes; }
+  ArrayRef<SourceLoc> getPossibleCommentLocs() const {
+    return PossibleCommentLocs;
+  }
+
+  /// Add an AST node to print.
+  void addNode(ASTNode Node) {
+    // Note we skip vars as they'll be printed as a part of their
+    // PatternBindingDecl.
+    if (!Node.isDecl(DeclKind::Var))
+      Nodes.push_back(Node);
+  }
+
+  /// Add a SourceLoc which may have a preceding comment attached. If so, the
+  /// comment will be printed out at the appropriate location.
+  void addPossibleCommentLoc(SourceLoc Loc) {
+    if (Loc.isValid())
+      PossibleCommentLocs.push_back(Loc);
+  }
+
+  /// Add all the nodes in the brace statement to the list of nodes to print.
+  /// This should be preferred over adding the nodes manually as it picks up the
+  /// end location of the brace statement as a possible comment loc, ensuring
+  /// that we print any trailing comments in the brace statement.
+  void addNodesInBraceStmt(BraceStmt *Brace) {
+    for (auto Node : Brace->getElements())
+      addNode(Node);
+
+    // Ignore the end locations of implicit braces, as they're likely bogus.
+    // e.g for a case statement, the r-brace loc points to the last token of the
+    // last node in the body.
+    if (!Brace->isImplicit())
+      addPossibleCommentLoc(Brace->getRBraceLoc());
+  }
+
+  /// Add the nodes and comment locs from another NodesToPrint.
+  void addNodes(NodesToPrint OtherNodes) {
```

Comment:
I'd probably prefer passing in a `const NodesToPrint &` here since we're not avoiding the copy anyway (ditto for the other similar methods).

Reply:
Talked with Ben offline, going to leave this as-is for now